### DISCUSSION TITLE
Why can't I play the same sound file twice on Android?

### DISCUSSION THREAD

**Discussion by jkloip** [flet-dev/flet#2925] [Q&A] [answered]
### Question

I can't play the same sound file repeatedly (using buttons) on Android. This problem doesn't exist on Windows. I don't know how I should write the code or make any settings?

### Code sample

```python
import flet as ft
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder"), center_title=True)

    path = "test-audio-file.wav" # a recording file
    url_1="MetroSVT.mp3"

    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)

    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=url_1,
        autoplay=False,
        volume=6,
        balance=0,
    )
    page.overlay.append(audio1)

    def handle_start_play(e):
        audio1.src=path
        audio1.update()
        audio1.play()

    await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play1", on_click=lambda _: audio1.play()),
        ft.ElevatedButton("Play2", on_click=handle_start_play),
    )

ft.app(target=main)
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by ndonkoHenri** [comment]
Can you try adding `page.on_error` ?
```py
page.on_error = lambda e: print(e.data)
```

**Reply by jkloip** [reply]
Okay, I'll test it later,thanks

**Reply by jkloip** [reply]
After I added page.on_error = lambda e: print(e.data) to the program, packaged the program into an apk and installed it on the Android tablet, no error message appeared when the same sound file was played for the second time, but the The second playback still produced no sound. This is a very strange phenomenon. I used the original example on the website to install the apk and had the same result. The same sound file had no sound at all when played for the second time.

**Reply by jkloip** [reply]
I use different file names for the same sound file and use different buttons to execute it. After I press "Play1" to play the sound, and then press "Play1" again, the sound file cannot play the sound again. But when I press "Play2" to play the sound, and then press "Play1" again, the sound file of "Play1" can be played perfectly. The program code is as follows, package it into apk format and install it on the android tablet.

import flet as ft
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder"), center_title=True)
    page.on_error = lambda e: print(e.data)

    path = "talk/passenger.m4a" # a recording file
    url_1="talk/MetroMLT1.mp3"
    url_2="talk/MetroMLT2.mp3"

    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)

    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=path,
        autoplay=False,
        volume=6,
        balance=0,
    )
    page.overlay.append(audio1)

    def handle_start_play1(e):
        audio1.src=" "
        audio1.update()
        audio1.src=url_1
        audio1.update()
        audio1.play()

    def handle_start_play2(e):
        audio1.src=" "
        audio1.update()
        audio1.src=url_2
        audio1.update()
        audio1.play()

    await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play1", on_click=handle_start_play1),
        ft.ElevatedButton("Play2", on_click=handle_start_play2),
    )

ft.app(target=main)

**Reply by jkloip** [reply]
I tried another way of writing, but the result is still the same. I need to execute the two buttons in turn so that their respective voice files can be played continuously. Does anyone know how I can solve this problem? My execution environment is "Android"

import flet as ft
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder"), center_title=True)
    page.on_error = lambda e: print(e.data)

    path = "talk/passenger.m4a" # a recording file
    url_1= "talk/MetroMLT1.mp3"
    url_2= "talk/staff_to_passenger.mp3"

    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)

    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=" ",
        autoplay=False,
        volume=6,
        balance=0,
    )
    page.overlay.append(audio1)

    def test1():
        audio1.src=None
        audio1.update()
        audio1.src=url_1
        audio1.update()

    def test2():
        audio1.src=None
        audio1.update()
        audio1.src=url_2
        audio1.update()

    def handle_start_play1(e):
        audio1.clean()
        test1()
        audio1.play()

    def handle_start_play2(e):
        test2()
        audio1.play()

    await page.add_async(
        #ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        #ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play1", on_click=handle_start_play1),
        ft.ElevatedButton("Play2", on_click=handle_start_play2),
    )

ft.app(target=main)

**Comment by jkloip** [comment, marked as answer]
After trying it, I found a solution for your reference. Just set **"release_mode=ft.audio.ReleaseMode.STOP"**. At this time, the same audio file can be played repeatedly on the Android system, which is great. The program code is as follows:

import flet as ft
async def main(page: ft.Page):
    page.horizontal_alignment = ft.CrossAxisAlignment.CENTER
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder"), center_title=True)
    page.on_error = lambda e: print(e.data)

    path = "talk/passenger.m4a" 
    url_1= "talk/MetroMLT1.mp3"
    url_2= "talk/staff_to_passenger.mp3"

    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)
    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)

    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=" ",
        autoplay=False,
        volume=6,
        balance=0,
        release_mode=ft.audio.ReleaseMode.STOP
    )
    page.overlay.append(audio1)

    def test1():
        pass

    def test2():
        audio1.src=url_2
        audio1.update()

    def handle_start_play1(e):
        audio1.src=url_1
        audio1.update()
        audio1.play()

    def handle_start_play2(e):
        test2()
        audio1.play()

    await page.add_async(
        ft.ElevatedButton("Start Audio Recorder", on_click=handle_start_recording),
        ft.ElevatedButton("Stop Audio Recorder", on_click=handle_stop_recording),
        ft.ElevatedButton("Play1", on_click=handle_start_play1),
        ft.ElevatedButton("Play2", on_click=handle_start_play2),
    )

ft.app(target=main)

**Comment by JosueJBA** [comment]
@jkloip Thanks for your response.  Your solution works, but I have another problem.  "ReleaseMode.STOP" stops audio from another application.  I need to play the sound from my app and keep the sound from the other app (ex: YouTube Music).  Could you help me?

**Reply by jkloip** [reply]
I don't know if this is the information you want.

import flet as ft

async def main(page: ft.Page):
    page.appbar = ft.AppBar(title=ft.Text("Audio Recorder"), center_title=True)

    path="talk/test0501.mp3"
    url1= "https://luan.xyz/files/audio/ambient_c_motion.mp3"
    url2= "https://github.com/mdn/webaudio-examples/blob/main/audio-analyser/viper.mp3?raw=true"

    async def handle_start_recording(e):
        await audio_rec.start_recording_async(path)

    async def handle_stop_recording(e):
        output_path = await audio_rec.stop_recording_async()
        if page.web and output_path is not None:
            await page.launch_url_async(output_path)

    audio_rec = ft.AudioRecorder(
        audio_encoder=ft.AudioEncoder.WAV,
    )
    page.overlay.append(audio_rec)
    await page.update_async()

    audio1 = ft.Audio(
        src=" ",
        autoplay=False,
        volume=6,
        balance=0,
        release_mode=ft.audio.ReleaseMode.STOP
    )
    page.overlay.append(audio1)

    audio2 = ft.Audio(
        src=" ",
        autoplay=False,
        volume=6,
        balance=0,
        release_mode=ft.audio.ReleaseMode.LOOP
    )
    page.overlay.append(audio2)

    def handle_start_play1(e):
        audio1.src=url1
        audio1.update()
        audio1.play()

    def handle_start_play2(e):
        audio2.src=url2
        audio2.update()
        audio2.play()

    sample_media = [
        ft.VideoMedia(
            "https://user-images.githubusercontent.com/28951144/229373720-14d69157-1a56-4a78-a2f4-d7a134d7c3e9.mp4"
        ),
    ]

    await page.add_async(
        video := ft.Video(
            expand=True,
            playlist=sample_media[0:2],
            playlist_mode=ft.PlaylistMode.LOOP,
            fill_color=ft.colors.BLUE_400,
            aspect_ratio=16/9,
            volume=100,
            autoplay=False,
            filter_quality=ft.FilterQuality.HIGH,
            muted=False,
            on_loaded=lambda e: print("Video loaded successfully!"),
            on_enter_fullscreen=lambda e: print("Video entered fullscreen!"),
            on_exit_fullscreen=lambda e: print("Video exited fullscreen!"),
        ),
        ft.ElevatedButton("Play1", on_click=handle_start_play1),
        ft.ElevatedButton("Play2", on_click=handle_start_play2),
        ft.ElevatedButton("Play1-STOP", on_click=lambda _: audio1.pause()),
    )

ft.app(target=main)

**Reply by JosueJBA** [reply]
Thanks.  Did not work.  "LOOP" mode only runs once.  STOP mode stops the audio from the other App.
